mater/lib/src/stores/blockstore.rs

```diff
@@ -1,370 +1,235 @@
+
+        let v1_payload_len = v1_header.encoded_len() as u64 + total_block_encoded_len;
+
+        v2::Header::new(
+            false,
+            v2::Header::SIZE,
+            v1_payload_len,
+            v2::Header::SIZE + v1_payload_len,
+        )
+    }
+}
+
+impl Blockwriter<Cursor<Vec<u8>>> {
+    /// Creates an in-memory [`Blockwriter`].
+    pub fn in_memory() -> Self {
         Self {
-            root: None,
-            blocks: IndexMap::new(),
-            indexed: HashSet::new(),
-            chunk_size: chunk_size.unwrap_or(DEFAULT_BLOCK_SIZE),
-            tree_width: tree_width.unwrap_or(DEFAULT_TREE_WIDTH),
+            writer: v2::Writer::new(Cursor::new(vec![])),
+            index: HashMap::new(),
+            roots: Vec::with_capacity(1),
+            started: false,
         }
     }
+}
 
-    /// Fully read the contents of an arbitrary `reader` into the [`Blockstore`],
-    /// converting the contents into a CARv2 file.
-    pub async fn read<R>(&mut self, reader: R) -> Result<(), Error>
+impl<W> Blockwriter<W>
+where
+    W: AsyncWrite + AsyncSeek + Unpin,
+{
+    /// Writes the contents from `source`, adding a new root to [`Blockwriter`].
+    pub async fn write_from<S>(&mut self, source: S) -> Result<(), Error>
     where
-        R: AsyncRead + Unpin,
+        S: AsyncRead + Unpin,
     {
-        let chunks = ReaderStream::with_capacity(reader, self.chunk_size);
-
-        // The `stream -> pin -> peekable` combo instead of `stream -> peekable -> pin` feels weird
-        // but it has to do with two things:
-        // - The fact that the stream can be self-referential:
-        //   https://users.rust-lang.org/t/why-is-pin-mut-needed-for-iteration-of-async-stream/51107
-        // - Using a tokio_stream::Peekable instead of futures::Peekable, they differ on who is required to be pinned
-        //  - tokio_stream::Peekable::peek(&mut self)
-        //    https://github.com/tokio-rs/tokio/blob/14c17fc09656a30230177b600bacceb9db33e942/tokio-stream/src/stream_ext/peekable.rs#L26-L37
-        //  - futures::Peekable::peek(self: Pin<&mut Self>)
-        //    https://github.com/rust-lang/futures-rs/blob/c507ff833728e2979cf5519fc931ea97308ec876/futures-util/src/stream/stream/peek.rs#L38-L40
-        let tree = stream_balanced_tree(chunks, self.tree_width);
-        tokio::pin!(tree);
-        let mut tree = tree.peekable();
-
-        while let Some(block) = tree.next().await {
-            let (cid, bytes) = block?;
-            self.insert(cid, bytes, true);
-
-            // If the stream is exhausted, we know the current block is the root
-            if tree.peek().await.is_none() {
-                // The root should always be indexed, there's no official spec saying it should though, it just makes sense.
-                // So, if the insert line is changed, the root should be placed in the `indexed` structure here
-                self.root = Some(cid);
-            }
+        if !self.started {
```

Comment:
What's up with this started stuff? Why would we support calling this function twice?

Reply:
Adding two files to the CAR. It creates two roots, even though that specific case is not fully supported this is here because when extending the API it will be necessary, though with a few adjustments.